File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        lVal = min(p.val, q.val)
        rVal = max(p.val, q.val)
        return self.findAncestor(root, lVal, rVal)

    def findAncestor(self, node: TreeNode, l: int, r: int) -> TreeNode:
        currV = node.val

        if currV == l or currV == r:
            return node
        
        if l < currV and currV < r:
            return node

        if currV > r:
            return self.findAncestor(node.left, l, r)
        
        return self.findAncestor(node.right, l, r)

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {    
    public bool IsSubtree(TreeNode root, TreeNode subRoot) {
        
        if (subRoot == null) {
            return true;
        }
        if (root == null) {
            return false;
        }

        if (SameTree(root, subRoot)) {
            return true;
        }
        return IsSubtree(root.left, subRoot) || 
               IsSubtree(root.right, subRoot);
    }

    public bool SameTree(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) {
            return true;
        }
        if (root != null && subRoot != null && root.val == subRoot.val) {
            return SameTree(root.left, subRoot.left) && 
                   SameTree(root.right, subRoot.right);
        }
        return false;
    }
}
